x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/Sample.java

```diff
@@ -0,0 +1,164 @@
+import org.elasticsearch.xpack.esql.expression.function.FunctionType;
+import org.elasticsearch.xpack.esql.expression.function.Param;
+import org.elasticsearch.xpack.esql.planner.PlannerUtils;
+import org.elasticsearch.xpack.esql.planner.ToAggregator;
+
+import java.io.IOException;
+import java.util.List;
+
+import static org.elasticsearch.common.logging.LoggerMessageFormat.format;
+
+public class Sample extends AggregateFunction implements ToAggregator {
+    public static final NamedWriteableRegistry.Entry ENTRY = new NamedWriteableRegistry.Entry(Expression.class, "Sample", Sample::new);
+
+    @FunctionInfo(
+        returnType = {
+            "boolean",
+            "cartesian_point",
+            "cartesian_shape",
+            "date",
+            "date_nanos",
+            "double",
+            "geo_point",
+            "geo_shape",
+            "integer",
+            "ip",
+            "keyword",
+            "long",
+            "version" },
+        description = "Collects sample values for a field.",
+        type = FunctionType.AGGREGATE,
+        examples = @Example(file = "stats_sample", tag = "doc")
+    )
+    public Sample(
+        Source source,
+        @Param(
+            name = "field",
+            type = {
+                "boolean",
+                "cartesian_point",
+                "cartesian_shape",
+                "date",
+                "date_nanos",
+                "double",
+                "geo_point",
+                "geo_shape",
+                "integer",
+                "ip",
+                "keyword",
+                "long",
+                "text",
+                "version" },
+            description = "The field to collect sample values for."
+        ) Expression field,
+        @Param(name = "limit", type = { "integer" }, description = "The maximum number of values to collect.") Expression limit
+    ) {
+        this(source, field, Literal.TRUE, limit);
+    }
+
+    public Sample(Source source, Expression field, Expression filter, Expression limit) {
+        this(source, field, filter, limit, new Literal(Source.EMPTY, Randomness.get().nextLong(), DataType.LONG));
+    }
+
+    /**
+     * The query "FROM data | STATS s1=SAMPLE(x,N), s2=SAMPLE(x,N)" should give two different
+     * samples of size N. The uuid is used to ensure that the optimizer does not optimize both
+     * expressions to one, resulting in identical samples.
+     */
+    public Sample(Source source, Expression field, Expression filter, Expression limit, Expression uuid) {
+        super(source, field, filter, List.of(limit, uuid));
+    }
+
+    private Sample(StreamInput in) throws IOException {
+        super(in);
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+        int limit = (int) limitField().fold(FoldContext.small());
```

Comment:
Nit: Can you use `limitValue()` here? Minor thing, but we're thinking about removing those `FoldContext.small()` usages (There's not a "clear" solution yet), and the more centralized, the better